src/mpi/misc/utils.c

```diff
@@ -40,43 +44,38 @@ int MPIR_Localcopy(const void *sendbuf, MPI_Aint sendcount, MPI_Datatype sendtyp
         copy_sz = sdata_sz;
 
     /* Builtin types is the common case; optimize for it */
-    if ((HANDLE_IS_BUILTIN(sendtype)) && HANDLE_IS_BUILTIN(recvtype)) {
-        MPIR_Memcpy(recvbuf, sendbuf, copy_sz);
-        goto fn_exit;
-    }
-
     MPIR_Datatype_iscontig(sendtype, &sendtype_iscontig);
     MPIR_Datatype_iscontig(recvtype, &recvtype_iscontig);
 
     MPIR_Type_get_true_extent_impl(sendtype, &sendtype_true_lb, &true_extent);
     MPIR_Type_get_true_extent_impl(recvtype, &recvtype_true_lb, &true_extent);
 
-    if (sendtype_iscontig && recvtype_iscontig) {
-#if defined(HAVE_ERROR_CHECKING)
-        MPIR_ERR_CHKMEMCPYANDJUMP(mpi_errno,
-                                  ((char *) recvbuf + recvtype_true_lb),
-                                  ((char *) sendbuf + sendtype_true_lb), copy_sz);
-#endif
-        MPIR_Memcpy(((char *) recvbuf + recvtype_true_lb),
-                    ((char *) sendbuf + sendtype_true_lb), copy_sz);
+    if (recvtype_iscontig) {
+        MPI_Aint actual_pack_bytes;
+        MPIR_Typerep_pack(sendbuf, sendcount, sendtype, 0, (char *) recvbuf + recvtype_true_lb,
+                          copy_sz, &actual_pack_bytes);
+        MPIR_ERR_CHKANDJUMP(actual_pack_bytes != copy_sz, mpi_errno, MPI_ERR_TYPE,
+                            "**dtypemismatch");
```

Comment:
Why is this branch moved up?  It's an unnecessary change.

Reply:
Right